### PR TITLE
Feature/7086 statistic data for vaccinations

### PR DESCRIPTION
**Pull request by 30mar**
## Description

- Add the new Vaccination Statistics Cards
- Unit and UITests

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7086

# video
https://user-images.githubusercontent.com/15270737/121706892-57378300-cad6-11eb-9da8-e1d26a4bb7a0.MP4

### CODE REVIEW

**Review by NataliaLemmerth — APPROVED** [review]
UA approved

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift [hunk 1]
```diff
@@ -19,13 +19,28 @@ extension SAP_Internal_Stats_KeyFigure {
...
 
-			return String(format: AppStrings.Statistics.Card.million, formattedNumber)
+			formattedString = String(format: AppStrings.Statistics.Card.million, formattedNumber)
 		} else {
 			let decimals = max(0, Int(self.decimals))
 			numberFormatter.minimumFractionDigits = Int(decimals)
 			numberFormatter.maximumFractionDigits = Int(decimals)
 
-			return numberFormatter.string(from: NSNumber(value: value))
+			formattedString = numberFormatter.string(from: NSNumber(value: value))
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift:28]: can we move the 10 mio stuff out to an extra method to make thing better readable?

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift [hunk 2]
```diff
@@ -19,13 +19,28 @@ extension SAP_Internal_Stats_KeyFigure {
...
+		case .atLeastOneVaccinatedPerson, .fullyVaccinatedPeople:
+			guard let valueString = formattedString else {
+				return formattedString
+			}
+			return valueString + AppStrings.Statistics.percent
+		case .infections, .incidence, .keySubmissions, .reproductionNumber, .appliedVaccinationsDoseRates:
+			return formattedString
+		case .none:
+			Log.info("Statistics card ID \(cardId) is not supported", log: .ui)
+			return formattedString
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift:43]: this is basically the same case as line 30 & 31
better call the method `formattedValue` with parameter `HomeStatisticsCard` so you can remove line 30 & 31 checks?

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift [hunk 3]
```diff
@@ -19,13 +19,28 @@ extension SAP_Internal_Stats_KeyFigure {
...
+		}
+		guard let cardId = cardRawValue else {
+			return formattedString
+		}
+		switch HomeStatisticsCard(rawValue: cardId) {
+		case .atLeastOneVaccinatedPerson, .fullyVaccinatedPeople:
+			guard let valueString = formattedString else {
+				return formattedString
+			}
+			return valueString + AppStrings.Statistics.percent
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift:38]: Would be better to use `numberFormatter.numberStyle = .percent`. Maybe we could just set `formattedValue(inPercent: Bool = false)` and handle the logic which card needs which in the `HomeStatisticsCardViewModel`?

**Review by nickguendling — CHANGES_REQUESTED** [review]

**Review by KaiTeuber — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/__tests__/HomeStatisticsCardViewModelTests.swift [hunk 4]
```diff
@@ -478,7 +698,16 @@ class HomeStatisticsCardViewModelTests: CWATestCase {
 
 				switch rank {
 				case .primary:
-					XCTAssertEqual(viewModel.primaryValue, expectedString)
+					switch HomeStatisticsCard(rawValue: id) {
+					case .atLeastOneVaccinatedPerson, .fullyVaccinatedPeople:
+						if inPercent {
+							XCTAssertEqual(viewModel.primaryValue, expectedString + AppStrings.Statistics.percent)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/__tests__/HomeStatisticsCardViewModelTests.swift:704]:
```suggestion
							XCTAssertEqual(viewModel.primaryValue, expectedString + "%")
```

For the tests you can use the percent symbol directly and I think then you can throw away the app string and the localizable string.

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift [hunk 5]
```diff
@@ -29,6 +28,14 @@ extension SAP_Internal_Stats_KeyFigure {
 		}
 	}
 
+	var formattedValueWithPercent: String? {
+		let numberFormatter = NumberFormatter()
+		numberFormatter.numberStyle = .percent
+		let decimals = max(0, Int(self.decimals))
+		numberFormatter.maximumFractionDigits = Int(decimals)
+		return numberFormatter.string(from: NSNumber(value: value / 100))?.filter({ !$0.isWhitespace })
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift:36]: Is the value that we get not already a percent value? Do we need to divide it by 100?

**Review by nickguendling — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift (hunk 5 above)
Comment (30mar) [reply src/xcode/ENA/ENA/Source/Scenes/Home/Cells/Statistics/Extensions/SAP_Internal_Stats_KeyFigure+Formatting.swift:36]: no the Formatted deals with the value of 1 as 100%  so if you pass 4 as a number it is considered 400% that is why i divide first

**Review by nickguendling — APPROVED** [review]

**Pull request merged**